lib/xlsxtream/worksheet.rb

```diff
@@ -8,11 +8,25 @@ def initialize(io, options = {})
       @io = io
       @rownum = 1
       @options = options
+      @sheetdata_written = false
 
       write_header
     end
 
+    # If you want to specify custom column widths, do so using this
+    # method. See Xlsstream::Columns#initialize for parameter details.
+    # This MUST be called before #add_row or #<<.
```

Comment:
Also if the order is fixed, I think it would be better to add optional options to the `#write_worksheet` method, that can be passed through to the `Worksheet.new` options, eg.:

```ruby
xlsx.write_worksheet 'Sheet1', columns: [width: 12, width: 40] do |sheet|
end
```

That would make it impossible to insert the `<cols/>` in the middle of the `<sheetData/>` and would avoid the `@sheetdata_written` checks.

Reply:
Ooops, I forgot the curly braces in the columns array, but you get the idea 😄

Reply:
http://www.officeopenxml.com/SScontentOverview.php

...says:

> The number and sizes of the columns of the grid are defined within a `<cols>`. And then the core data of the worksheet follows within the `<sheetData>` element

"Follows" implies that `<cols>` must be after. It definitely cannot be written within the sheetdata, as that's what led to the 'written' flag.

The idea of going back to the drawing board on the PR as-written and moving things around so that the `write_worksheet` method takes a new option does seem a lot cleaner at this point.

Reply:
Sounds good to me!